dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp

```diff
@@ -389,9 +441,293 @@ SegmentPtr DeltaMergeStore::segmentMerge(
+    const DMFiles & dm_files,
+    const String & source_segment_info)
+{
+    // 1. Acquire a snapshot for PageStorage, and keep the snapshot until index is built.
+    // This helps keep DMFile valid during the index build process.
+    // We don't acquire a snapshot from the source_segment, because the source_segment
+    // may be abandoned at this moment.
+    //
+    // Note that we cannot simply skip the index building when seg is not valid any more,
+    // because segL and segR is still referencing them, consider this case:
+    //   1. seg=PhysicalSplit
+    //   2. Add CreateStableIndex(seg) to ThreadPool
+    //   3. segL, segR=LogicalSplit(seg)
+    //   4. CreateStableIndex(seg)
+
+    auto storage_snapshot = std::make_shared<StorageSnapshot>(
+        *dm_context.storage_pool,
+        dm_context.getReadLimiter(),
+        dm_context.tracing_id,
+        /*snapshot_read*/ true);
+
+    RUNTIME_CHECK(dm_files.size() == 1); // size > 1 is currently not supported.
+    const auto & dm_file = dm_files[0];
+
+    // 2. Check whether the DMFile has been referenced by any valid segment.
+    {
+        std::shared_lock lock(read_write_mutex);
+        auto segment_ids = dmfile_id_to_segment_ids.get(dm_file->fileId());
+        if (segment_ids.empty())
+        {
+            LOG_DEBUG(
+                log,
+                "EnsureStableIndex - Give up because no segment to update, source_segment={}",
+                source_segment_info);
+            return;
+        }
+    }
+
+    LOG_INFO(
+        log,
+        "EnsureStableIndex - Begin building index, dm_files={} source_segment={}",
+        DMFile::info(dm_files),
+        source_segment_info);
+
+    // 2. Build the index.
+    DMFileIndexWriter iw(DMFileIndexWriter::Options{
+        .path_pool = path_pool,
+        .index_infos = index_info,
+        .dm_files = dm_files,
+        .dm_context = dm_context,
+    });
+    auto new_dmfiles = iw.build();
+    RUNTIME_CHECK(!new_dmfiles.empty());
+
+    LOG_INFO(
+        log,
+        "EnsureStableIndex - Finish building index, dm_files={} source_segment={}",
+        DMFile::info(dm_files),
+        source_segment_info);
+
+    // 3. Update the meta version of the segments to the latest one.
+    // To avoid logical split between step 2 and 3, get lastest segments to update again.
+    // If TiFlash crashes during updating the meta version, some segments' meta are updated and some are not.
+    // So after TiFlash restarts, we will update meta versions to latest versions again.
+    {
+        // We must acquire a single lock when updating multiple segments.
+        // Otherwise we may miss new segments.
+        std::unique_lock lock(read_write_mutex);
+        auto segment_ids = dmfile_id_to_segment_ids.get(dm_file->fileId());
+        for (const auto & seg_id : segment_ids)
+        {
+            auto segment = id_to_segment[seg_id];
+            auto new_segment = segmentUpdateMeta(lock, dm_context, segment, new_dmfiles);
```

Comment:
In segmentUpdateMeta, we will acquire a lock for `using Lock = std::unique_lock<std::recursive_mutex>;` of a DeltaValueSpace for a Segment, while we are holding also a unique_lock of DM.
I wonder if the lock is too grain? AFAIK, the segment lock is only used in `replaceStableMetaVersion`. So I wonder if we could do `replaceStableMetaVersion` without DM lock?

Reply:
Accquiring DM lock to make sure segment will not be abandoned by other thread.